codex/blockexchange/engine/advertiser.nim

```diff
@@ -0,0 +1,173 @@
+    discovery*: Discovery                                        # Discovery interface
+
+    advertiserRunning*: bool                                     # Indicates if discovery is running
+    concurrentAdvReqs: int                                       # Concurrent advertise requests
+
+    advertiseLocalStoreLoop*: Future[void]                                 # Advertise loop task handle
+    advertiseQueue*: AsyncQueue[Cid]                             # Advertise queue
+    advertiseTasks*: seq[Future[void]]                           # Advertise tasks
+
+    advertiseLocalStoreLoopSleep: Duration                                 # Advertise loop sleep
+    inFlightAdvReqs*: Table[Cid, Future[void]]                   # Inflight advertise requests
+
+proc addCidToQueue(b: Advertiser, cid: Cid) {.async.} =
+  if cid notin b.advertiseQueue:
+    await b.advertiseQueue.put(cid)
+    trace "Advertising", cid
+
+proc advertiseBlock(b: Advertiser, cid: Cid) {.async.} =
+  without isM =? cid.isManifest, err:
+    warn "Unable to determine if cid is manifest"
+    return
+
+  if isM:
+    without blk =? await b.localStore.getBlock(cid), err:
+      error "Error retrieving manifest block", cid, err = err.msg
+      return
+
+    without manifest =? Manifest.decode(blk), err:
+      error "Unable to decode as manifest", err = err.msg
+      return
+
+    # announce manifest cid and tree cid
+    await b.addCidToQueue(cid)
+    await b.addCidToQueue(manifest.treeCid)
+
+proc advertiseLocalStoreLoop(b: Advertiser) {.async.} =
+  while b.advertiserRunning:
+    if cids =? await b.localStore.listBlocks(blockType = BlockType.Manifest):
+      trace "Advertiser begins iterating blocks..."
+      for c in cids:
+        if cid =? await c:
+          await b.advertiseBlock(cid)
+      trace "Advertiser iterating blocks finished."
+
+    await sleepAsync(b.advertiseLocalStoreLoopSleep)
+
+  info "Exiting advertise task loop"
+
+proc processQueueLoop(b: Advertiser) {.async.} =
+  while b.advertiserRunning:
+    try:
+      let
+        cid = await b.advertiseQueue.get()
+
+      if cid in b.inFlightAdvReqs:
+        continue
+
+      try:
+        let
+          request = b.discovery.provide(cid)
+
+        b.inFlightAdvReqs[cid] = request
+        codexInflightAdvertise.set(b.inFlightAdvReqs.len.int64)
+        await request
+
+      finally:
+        b.inFlightAdvReqs.del(cid)
+        codexInflightAdvertise.set(b.inFlightAdvReqs.len.int64)
+    except CancelledError:
+      trace "Advertise task cancelled"
+      return
+    except CatchableError as exc:
+      warn "Exception in advertise task runner", exc = exc.msg
+
+  info "Exiting advertise task runner"
+
+proc start*(b: Advertiser) {.async.} =
+  ## Start the advertiser
+  ##
+
+  trace "Advertiser start"
+
+  proc onBlock(cid: Cid) {.async.} = 
+    await b.advertiseBlock(cid)
+
+  b.localStore.onBlockStored = onBlock.some
```

Comment:
This is error-prone: if someone has already installed a callback, you will simply overwrite it. I'd change the `onBlockStored` assignment so that if it's already set, this results in an error.

Reply:
Will do. (I suspect there are plenty of places in the codebase where this exact thing could happen.)